libevm/VMSIMD.cpp

```diff
@@ -0,0 +1,729 @@
+#define EVALXOPU(OP, _b) EVALXOP(OP, uint8_t, uint16_t, uint32_t, uint64_t, _b)
+#define EVALXOP(OP, T8, T16, T32, T64, _b) \
+{ \
+	uint8_t const t = (_b) & 0xf; \
+	m_SPP[0] = 0; \
+	switch (t) \
+	{ \
+	case 0: \
+		for (int i = 0; i < 32; ++i) \
+			v8x32(m_SPP[0])[i]  = (uint8_t) OP((T8) v8x32(m_SP[0])[i],  (T8) v8x32(m_SP[1])[i]); \
+		break; \
+	case 1: \
+		for (int i = 0; i < 16; ++i) \
+			v16x16(m_SPP[0])[i] = (uint16_t)OP((T16)v16x16(m_SP[0])[i], (T16)v16x16(m_SP[1])[i]); \
+		break; \
+	case 2: \
+		for (int i = 0; i < 8; ++i) \
+			v32x8(m_SPP[0])[i]  = (uint32_t)OP((T32)v32x8(m_SP[0])[i],  (T32)v32x8(m_SP[1])[i]); \
+		break; \
+	case 3: \
+		for (int i = 0; i < 4; ++i) \
+			v64x4(m_SPP[0])[i]  = (uint64_t)OP((T64)v64x4(m_SP[0])[i],  (T64)v64x4(m_SP[1])[i]); \
+		break; \
+	default: throwBadInstruction(); \
+	} \
+}
+#define ADD( x1, x2) ((x1) + (x2))
+#define MUL( x1, x2) ((x1) * (x2))
+#define SUB( x1, x2) ((x1) - (x2))
+#define DIV( x1, x2) ((x1) / (x2))
+#define MOD( x1, x2) ((x1) % (x2))
+#define LT(  x1, x2) ((x1) < (x2))
+#define GT(  x1, x2) ((x1) > (x2))
+#define EQ(  x1, x2) ((x1) == (x2))
+#define ZERO(x1, x2) ((x1) == 0)
+#define AND( x1, x2) ((x1) & (x2))
+#define OR(  x1, x2) ((x1) | (x2))
+#define XOR( x1, x2) ((x1) ^ (x2))
+#define NOT( x1, x2) (~x1)
+#define SHR( x1, x2) ((x1) >> (x2))
+#define SHL( x1, x2) ((x1) << (x2))
+#define ROL( x1, x2) (((x1) << x2)|((x1) >> (sizeof(x1) * 8 - (x2))))
+#define ROR( x1, x2) (((x1) >> x2)|((x1) << (sizeof(x1) * 8 - (x2))))
+
+void VM::xadd (uint8_t _b) { EVALXOPU(ADD, _b); }
+void VM::xmul (uint8_t _b) { EVALXOPU(MUL, _b); }
+void VM::xsub (uint8_t _b) { EVALXOPU(SUB, _b); }
+void VM::xdiv (uint8_t _b) { EVALXOPU(DIV, _b); }
+void VM::xsdiv(uint8_t _b) { EVALXOPS(DIV, _b); }
+void VM::xmod (uint8_t _b) { EVALXOPU(MOD, _b); }
+void VM::xsmod(uint8_t _b) { EVALXOPS(MOD, _b); }
+void VM::xlt  (uint8_t _b) { EVALXOPU(LT,  _b); }
+void VM::xslt (uint8_t _b) { EVALXOPS(LT,  _b); }
+void VM::xgt  (uint8_t _b) { EVALXOPU(GT,  _b); }
+void VM::xsgt (uint8_t _b) { EVALXOPS(GT,  _b); }
+void VM::xeq  (uint8_t _b) { EVALXOPU(EQ,  _b); }
+void VM::xzero(uint8_t _b) { EVALXOPU(ZERO,_b); }
+void VM::xand (uint8_t _b) { EVALXOPU(AND, _b); }
+void VM::xoor (uint8_t _b) { EVALXOPU(OR,  _b); }
+void VM::xxor (uint8_t _b) { EVALXOPU(XOR, _b); }
+void VM::xnot (uint8_t _b) { EVALXOPU(NOT, _b); }
+void VM::xshr (uint8_t _b) { EVALXOPU(SHR, _b); }
+void VM::xsar (uint8_t _b) { EVALXOPS(SHR, _b); }
+void VM::xshl (uint8_t _b) { EVALXOPU(SHL, _b); }
+void VM::xrol (uint8_t _b) { EVALXOPU(ROL, _b); }
+void VM::xror (uint8_t _b) { EVALXOPU(ROR, _b); }
+
+inline uint8_t pow2N(uint8_t n)
+{
+	static uint8_t exp[6] = { 1, 2, 4, 8, 16, 32 };
+	return exp[n];
+}
+
+inline uint8_t nElem(uint8_t _b)
+{
+	return pow2N((_b) & 0xf);
+}
+
+inline uint8_t elemType(uint8_t _b)
+{
+	return (_b) >> 4;
+}
+
+// in must be by reference because it is really just memory for a vector
```

Comment:
was that comment about output param? there's no output param anymore, comment may be removed, I guess